packages/embed-jsx/src/lib/EmbedJsx.ts

```diff
@@ -0,0 +1,142 @@
+import { MessageEmbed, ColorResolvable } from 'discord.js';
+
+const EMBED_TYPES = ['embed', 'title', 'field', 'timestamp', 'footer', 'description', 'url', 'image'] as const;
+
+type EmbedInformation =
+	| TitleInformation
+	| FieldInformation
+	| TimestampInformation
+	| FooterInformation
+	| DescriptionInformation
+	| UrlInformation
+	| ImageInformation;
+
+type EmbedData = TitleData | FieldData | TimestampData | FooterData | DescriptionData | UrlData | ImageData | EmptyData;
+
+const enum TsxTypes {
+	Title,
+	Description,
+	Url,
+	Field,
+	Timestamp,
+	Footer,
+	Image,
+	Empty
+}
+
+// Received types
+type TitleInformation = [null, string];
+type DescriptionInformation = [null, string];
+type UrlInformation = [null, string];
+type FieldInformation = [{ title?: string; inline?: boolean }, string];
+type TimestampInformation = [null, number | string | Date | null];
+type FooterInformation = [null | { iconURL?: string }, string];
+type EmbedInitialInformation = [{ color: ColorResolvable } | null, ...EmbedData[]];
+type ImageInformation = [{ url: string }, null];
+
+// Returned types
+type TitleData = [TsxTypes.Title, string];
+type DescriptionData = [TsxTypes.Description, string];
+type UrlData = [TsxTypes.Url, string];
+type FieldData = [TsxTypes.Field, string, string, boolean];
+type TimestampData = [TsxTypes.Timestamp, Date];
+type FooterData = [TsxTypes.Footer, string, string?];
+type ImageData = [TsxTypes.Image, string];
+type EmptyData = [TsxTypes.Empty];
+
+/**
+ * The namespace to import for embed-jsx
+ *
+ * @example
+ * ```tsx
+ * import { EmbedJsx } from '@sapphire/embed-jsx';
+ *
+ * const embed = (
+ *   <embed color="RED">
+ * 	   <title>New Embed</title>
+ * 	   <description>Hello!</description>
+ *   </embed>
+ * )
+ * ```
+ */
+// eslint-disable-next-line @typescript-eslint/no-namespace
+export namespace EmbedJsx {
+	/**
+	 * The behind the scenes function that TS uses to turn JSX into proper JS
+	 * @param type The embed type being created
+	 * @param data The data emitted along with
+	 * @returns The embed defined with jsx
+	 */
+	export function make(type: typeof EMBED_TYPES[number], ...data: EmbedInformation | EmbedInitialInformation): EmbedData | MessageEmbed {
+		type = type.toLowerCase() as typeof EMBED_TYPES[number];
+		if (!EMBED_TYPES.includes(type)) throw new TypeError(`Invalid type passed, expected one of ${EMBED_TYPES.join(', ')}, got: ${type}`);
+		switch (type) {
+			case 'embed': {
+				const info = data as EmbedInitialInformation;
+				let embed = new MessageEmbed(info[0] ?? {});
+				info.shift();
```

Comment:
Can't you unite these? `info.shift() ?? {}`?

Reply:
will the `shift` still occur if the value is null?

Reply:
![image](https://user-images.githubusercontent.com/17960496/114314647-25209980-9b04-11eb-8bdb-b6231d37ba8b.png)

Reply:
Fixed in 4edaf9bc198e25e4d4222c33a1d7e5b891cb338a